### PR TITLE
Refactor SdkClient to a concrete class with delegate implementation

### PR DESCRIPTION
**Pull request by dbwiddis**
### Description

Refactors `SdkClient` interface to a concrete class containing a delegate implementation.  This allows it to be injected in the `createComponents()` extension point and available to `@Inject` constructors.

Uses Cluster Settings to determine which implementation is generated by the SdkClientFactory.

Adds SSL/TLS and Basic Auth to the remote client.

Adds an AWS OpenSearch Service client.
  
### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: plugin/src/main/java/org/opensearch/ml/sdkclient/SdkClientFactory.java [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+
+/**
+ * A class to create a {@link SdkClient} with implementation based on settings
+ */
+@Log4j2
+public class SdkClientFactory {
+
+    // TODO Move these settings keys to plugin getSettings implementation with proper prefixes
+    public static final String REMOTE_METADATA_TYPE = "REMOTE_METADATA_TYPE";
+    public static final String REMOTE_METADATA_ENDPOINT = "REMOTE_METADATA_ENDPOINT";
```
Comment (arjunkumargiri) [inline plugin/src/main/java/org/opensearch/ml/sdkclient/SdkClientFactory.java:51]: Along with prefix, lets ensure that it follows settings naming convention.

File: common/src/main/java/org/opensearch/sdk/SdkClientImpl.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+package org.opensearch.sdk;
+
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.Executor;
+
+public abstract class SdkClientImpl {
```
Comment (arjunkumargiri) [inline common/src/main/java/org/opensearch/sdk/SdkClientImpl.java:14]: This should be an interface instead of abstract class

Comment (arjunkumargiri) [reply common/src/main/java/org/opensearch/sdk/SdkClientImpl.java:14]: Rename to SDKClientInterface?

File: common/src/main/java/org/opensearch/sdk/SdkClientImpl.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+     * @param executor the executor to use for asynchronous execution
+     * @return A completion stage encapsulating the response or exception
+     */
+    public abstract CompletionStage<SearchDataObjectResponse> searchDataObjectAsync(SearchDataObjectRequest request, Executor executor);
+    
+    /**
+     * Get the underlying client implementation
+     * @return the client implementation
+     */
+    public abstract Object getImplementation();
```
Comment (arjunkumargiri) [inline common/src/main/java/org/opensearch/sdk/SdkClientImpl.java:64]: Why is this required? underlying implementation need not be exposed

File: plugin/src/main/java/org/opensearch/ml/sdkclient/SdkClientFactory.java (hunk 1 above)
Comment (dbwiddis) [reply plugin/src/main/java/org/opensearch/ml/sdkclient/SdkClientFactory.java:51]: The TODO commend does include "with proper prefixes" :)

File: common/src/main/java/org/opensearch/sdk/SdkClientImpl.java (hunk 3 above)
Comment (dbwiddis) [reply common/src/main/java/org/opensearch/sdk/SdkClientImpl.java:64]: At one point I thought I needed to add it for tests but then later added the "wrap" constructor for tests so it became obsolete.  It can go, I guess. I thought it might be useful.

File: common/src/main/java/org/opensearch/sdk/SdkClientImpl.java (hunk 2 above)
Comment (dbwiddis) [reply common/src/main/java/org/opensearch/sdk/SdkClientImpl.java]: SdkClientInterface comes from the Department of Redundancy Department. :)

I did SdkClientDelegate.

File: common/src/test/java/org/opensearch/sdk/SdkClientTests.java [hunk 4]
```diff
@@ -86,162 +87,163 @@ public CompletionStage<SearchDataObjectResponse> searchDataObjectAsync(SearchDat
                 return CompletableFuture.completedFuture(searchResponse);
             }
         });
+        sdkClient = new SdkClient(sdkClientImpl);
```
Comment (dhrubo-os) [inline common/src/test/java/org/opensearch/sdk/SdkClientTests.java:90]: I assume we can use `sdkClient` like this way in algorithms module as we can't import SDKClientFactory class to algorithm module?

File: plugin/src/main/java/org/opensearch/ml/plugin/MachineLearningPlugin.java [hunk 5]
```diff
@@ -654,7 +649,7 @@ public Collection<Object> createComponents(
             .getClusterSettings()
             .addSettingsUpdateConsumer(MLCommonsSettings.ML_COMMONS_RAG_PIPELINE_FEATURE_ENABLED, it -> ragSearchPipelineEnabled = it);
 
-        LocalClusterIndicesClient localClusterIndicesClient = new LocalClusterIndicesClient(client, xContentRegistry);
+        SdkClient sdkClient = SdkClientFactory.createSdkClient(client, xContentRegistry, settings);
```
Comment (dhrubo-os) [inline plugin/src/main/java/org/opensearch/ml/plugin/MachineLearningPlugin.java:652]: Cool, so that means we don't need to do ?

```
ModulesBuilder modules = new ModulesBuilder();
modules.add(new SdkClientModule(client, xContentRegistry));

 injector = modules.createInjector();

 // Get the injected SdkClient instance from the injector
 SdkClient sdkClient = injector.getInstance(SdkClient.class);

File: common/src/main/java/org/opensearch/sdk/SdkClientSettings.java [hunk 6]
```diff
@@ -0,0 +1,36 @@
+/*
```
Comment (dhrubo-os) [inline common/src/main/java/org/opensearch/sdk/SdkClientSettings.java:1]: I assume these settings will eventually move to core, right?

Otherwise we usually keep our settings in one place: https://github.com/opensearch-project/ml-commons/blob/0b9708b831f87ac6eb35b5c3039c61afe9425d26/plugin/src/main/java/org/opensearch/ml/settings/MLCommonsSettings.java#L17

File: plugin/src/main/java/org/opensearch/ml/plugin/MachineLearningPlugin.java (hunk 5 above)
Comment (dbwiddis) [reply plugin/src/main/java/org/opensearch/ml/plugin/MachineLearningPlugin.java:652]: Correct.  That created an injector that was different than the one OpenSearch processes with the list of objects from `createComponents()`.  But the clients still had their runtime class so those would still have been bound to the subclasses, and this really just made a new/different SdkClient available to the various other classes in this method.   We could have just instantiated an sdkClient without jumping through these hoops and gotten the same thing.

With the new code we have a concrete class `SdkClient` that is injected and available by its class key (which is what the separate module was doing) and hides the implementation as an internal field.   Not the most object-oriented way to do things, but it works around the injection headaches.

File: common/src/main/java/org/opensearch/sdk/SdkClientSettings.java (hunk 6 above)
Comment (dbwiddis) [reply common/src/main/java/org/opensearch/sdk/SdkClientSettings.java:1]: Yes, these are intended to be moved to wherever this whole package will eventually move (an OpenSearch module?  Another plugin?  The SDK for Java extensions repo? But it will not remain in ML Commons long term and those settings names will probably change when we relocate it.

File: common/src/test/java/org/opensearch/sdk/SdkClientTests.java (hunk 4 above)
Comment (dbwiddis) [reply common/src/test/java/org/opensearch/sdk/SdkClientTests.java:90]: Yes but you'd still need the LocalClusterIndicesClient.  One possibility is moving that default client to the `common` module since it doesn't need the Jackson stuff that the DDB Client does.

**Review by arjunkumargiri — APPROVED** [review]

File: common/src/test/java/org/opensearch/sdk/SdkClientTests.java (hunk 4 above)
Comment (dhrubo-os) [reply common/src/test/java/org/opensearch/sdk/SdkClientTests.java:90]: That could be a easy workaround. Do you want to address that in this PR? Or in another PR?

Comment (dbwiddis) [reply common/src/test/java/org/opensearch/sdk/SdkClientTests.java:90]: Probably another PR. I may make another attempt to solve the jackson problems.

**Review by dhrubo-os — APPROVED** [review]
